Source/WarpXEvolve.cpp

```diff
@@ -252,6 +222,173 @@ WarpX::EvolveEM (int numsteps)
     }
 }
 
+void
+WarpX::OneStep_nosub (Real cur_time)
+{
+    // Push particle from x^{n} to x^{n+1}
+    //               from p^{n-1/2} to p^{n+1/2}
+    // Deposit current j^{n+1/2}
+    // Deposit charge density rho^{n}
+#ifdef WARPX_USE_PY
+    if (warpx_py_particleinjection) warpx_py_particleinjection();
+    if (warpx_py_particlescraper) warpx_py_particlescraper();
+    if (warpx_py_beforedeposition) warpx_py_beforedeposition();
+#endif
+    PushParticlesandDepose(cur_time);
+#ifdef WARPX_USE_PY
+    if (warpx_py_afterdeposition) warpx_py_afterdeposition();
+#endif
+
+    SyncCurrent();
+
+    SyncRho(rho_fp, rho_cp);
+
+    // Push E and B from {n} to {n+1}
+    // (And update guard cells immediately afterwards)
+#ifdef WARPX_USE_PSATD
+    PushPSATD(dt[0]);
+    FillBoundaryE();
+    FillBoundaryB();
+#else
+    EvolveF(0.5*dt[0], DtType::FirstHalf);
+    FillBoundaryF();
+    EvolveB(0.5*dt[0]); // We now have B^{n+1/2}
+    FillBoundaryB();
+    EvolveE(dt[0]); // We now have E^{n+1}
+    FillBoundaryE();
+    EvolveF(0.5*dt[0], DtType::SecondHalf);
+    EvolveB(0.5*dt[0]); // We now have B^{n+1}
+    if (do_pml) {
+        DampPML();
+        FillBoundaryE();
+    }
+    FillBoundaryB();
+#endif
+}
+
+void
+WarpX::OneStep_sub1 (Real curtime)
+{
+    // TODO: we could save some charge depositions
```

Comment:
@WeiqunZhang Could you elaborate on the above?

Reply:
Every time we push particles, we compute charge density for t and t+dt.  Rho at t+dt could be potentially used as pre-push density for next push.